src/rules/no_inner_declarations.rs

```diff
@@ -0,0 +1,286 @@
+
+  fn visit_module_item(&mut self, item: &ast::ModuleItem, parent: &dyn Node) {
+    match item {
+      ast::ModuleItem::ModuleDecl(module_decl) => match module_decl {
+        ast::ModuleDecl::ExportDecl(decl_export) => {
+          self.check_decl(&decl_export.decl)
+        }
+        ast::ModuleDecl::ExportDefaultDecl(default_export) => {
+          if let ast::DefaultDecl::Fn(fn_expr) = &default_export.decl {
+            self.valid_decls.push(fn_expr.span());
+          }
+        }
+        _ => {}
+      },
+      ast::ModuleItem::Stmt(module_stmt) => {
+        if let ast::Stmt::Decl(decl) = module_stmt {
+          self.check_decl(decl)
+        }
+      }
+    }
+
+    swc_ecmascript::visit::visit_module_item(self, item, parent);
+  }
+
+  fn visit_function(&mut self, function: &ast::Function, parent: &dyn Node) {
+    if let Some(block) = &function.body {
+      for stmt in &block.stmts {
+        if let ast::Stmt::Decl(decl) = stmt {
+          self.check_decl(decl);
+        }
+      }
+    }
+    swc_ecmascript::visit::visit_function(self, function, parent);
+  }
+
+  fn visit_fn_decl(&mut self, fn_decl: &ast::FnDecl, parent: &dyn Node) {
+    if let Some(block) = &fn_decl.function.body {
+      for stmt in &block.stmts {
+        if let ast::Stmt::Decl(decl) = stmt {
+          self.check_decl(decl);
+        }
+      }
+    }
+    swc_ecmascript::visit::visit_fn_decl(self, fn_decl, parent);
+  }
+
+  fn visit_fn_expr(&mut self, fn_expr: &ast::FnExpr, parent: &dyn Node) {
+    if let Some(block) = &fn_expr.function.body {
+      for stmt in &block.stmts {
+        if let ast::Stmt::Decl(decl) = stmt {
+          self.check_decl(decl);
+        }
+      }
+    }
+    swc_ecmascript::visit::visit_fn_expr(self, fn_expr, parent);
+  }
+
+  fn visit_arrow_expr(
+    &mut self,
+    arrow_expr: &ast::ArrowExpr,
+    parent: &dyn Node,
+  ) {
+    if let ast::BlockStmtOrExpr::BlockStmt(block) = &arrow_expr.body {
+      for stmt in &block.stmts {
+        if let ast::Stmt::Decl(decl) = stmt {
+          self.check_decl(decl);
+        }
+      }
+    }
+    swc_ecmascript::visit::visit_arrow_expr(self, arrow_expr, parent);
+  }
+}
+
+struct NoInnerDeclarationsVisitor {
+  context: Arc<Context>,
+  valid_decls: Vec<Span>,
+  in_function: bool,
+}
+
+impl NoInnerDeclarationsVisitor {
+  fn new(context: Arc<Context>, valid_decls: Vec<Span>) -> Self {
+    Self {
+      context,
+      valid_decls,
+      in_function: false,
+    }
+  }
+}
+
+impl NoInnerDeclarationsVisitor {
+  fn add_diagnostic(&mut self, span: Span, kind: &str) {
+    let root = if self.in_function {
+      "function"
+    } else {
+      "module"
+    };
+
+    self.context.add_diagnostic(
+      span,
+      "no-inner-declarations",
+      &format!("Move {} declaration to {} root", kind, root),
+    );
+  }
+}
+
+impl Visit for NoInnerDeclarationsVisitor {
+  noop_visit_type!();
+
+  fn visit_function(&mut self, fn_: &ast::Function, parent: &dyn Node) {
```

Comment:
Seems that implementing `visit_arrow_expr` here is also necessary to check if it's inside function

Reply:
Ah yeah, good catch 👍

Reply:
Done